pkg/handler/v2/preset/preset.go

```diff
@@ -0,0 +1,368 @@
+		}
+	}
+
+	return nil
+}
+
+func DecodeCreatePreset(_ context.Context, r *http.Request) (interface{}, error) {
+	var req createPresetReq
+
+	req.ProviderName = mux.Vars(r)["provider_name"]
+	if err := json.NewDecoder(r.Body).Decode(&req.Body); err != nil {
+		return nil, err
+	}
+
+	return req, nil
+}
+
+// CreatePreset creates a preset for the selected provider and returns the name if successful, error otherwise
+func CreatePreset(presetsProvider provider.PresetProvider, userInfoGetter provider.UserInfoGetter) endpoint.Endpoint {
+	return func(ctx context.Context, request interface{}) (interface{}, error) {
+		req, ok := request.(createPresetReq)
+		if !ok {
+			return nil, errors.NewBadRequest("invalid request")
+		}
+
+		err := req.Validate()
+		if err != nil {
+			return nil, errors.NewBadRequest(err.Error())
+		}
+
+		userInfo, err := userInfoGetter(ctx, "")
+		if err != nil {
+			return nil, common.KubernetesErrorToHTTPError(err)
+		}
+
+		preset, err := presetsProvider.GetPreset(userInfo, req.Body.Name)
+		if err != nil && !k8serrors.IsNotFound(err) {
+			return nil, err
+		}
+
+		if k8serrors.IsNotFound(err) {
+			return presetsProvider.CreatePreset(userInfo, &req.Body)
+		}
+
+		if hasProvider, _ := preset.Spec.HasProvider(crdapiv1.ProviderType(req.ProviderName)); hasProvider {
+			return nil, fmt.Errorf("preset with name %s already exists for the provider %s", preset.Name, req.ProviderName)
+		}
+
+		preset = mergePresets(preset, &req.Body, crdapiv1.ProviderType(req.ProviderName))
+		return presetsProvider.UpdatePreset(userInfo, preset)
+	}
+}
+
+// updatePresetReq represents a request to update a preset
+// swagger:parameters updatePreset
+type updatePresetReq struct {
+	createPresetReq
+}
+
+// Validate validates updatePresetReq request
+func (r updatePresetReq) Validate() error {
+	return r.createPresetReq.Validate()
+}
+
+func DecodeUpdatePreset(_ context.Context, r *http.Request) (interface{}, error) {
+	var req updatePresetReq
+
+	req.ProviderName = mux.Vars(r)["provider_name"]
+	if err := json.NewDecoder(r.Body).Decode(&req.Body); err != nil {
+		return nil, err
+	}
+
+	return req, nil
+}
+
+// UpdatePreset updates a preset for the selected provider and returns the name if successful, error otherwise
+func UpdatePreset(presetsProvider provider.PresetProvider, userInfoGetter provider.UserInfoGetter) endpoint.Endpoint {
+	return func(ctx context.Context, request interface{}) (interface{}, error) {
+		req, ok := request.(updatePresetReq)
+		if !ok {
+			return nil, errors.NewBadRequest("invalid request")
+		}
+
+		err := req.Validate()
+		if err != nil {
+			return nil, errors.NewBadRequest(err.Error())
+		}
+
+		userInfo, err := userInfoGetter(ctx, "")
+		if err != nil {
+			return nil, common.KubernetesErrorToHTTPError(err)
+		}
+
+		preset, err := presetsProvider.GetPreset(userInfo, req.Body.Name)
+		if err != nil && !k8serrors.IsNotFound(err) {
```

Comment:
if not found then it will be notified, not needed to return a custom error below

Reply:
ahh... yes. I have to update the logic.